### PR TITLE
FOGL-3029:  support for configuration item validity

### PR DESCRIPTION
**Pull request by shariq1501**
Initial testing with dummy data [payload.zip](https://github.com/foglamp/foglamp-gui/files/3635971/payload.zip)

To create a category 
`curl -sX POST http://localhost:8081/foglamp/category -d @demo.json | jq`

### CODE REVIEW

**Review by praveen-garg — APPROVED** [review]

**Review by praveen-garg — CHANGES_REQUESTED** [review]
[demo.json.zip](https://github.com/foglamp/foglamp-gui/files/3640003/demo.json.zip)

Screenshot 2019-09-22 at 10 57 14 PM

`demo1` should always be editable. but that is always non-editable!

also try `validity=""`

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 1]
```diff
@@ -352,4 +346,52 @@ export class ViewConfigItemComponent implements OnInit, OnChanges, AfterViewChec
...
+        }
+      }
+
+      for (const k in data) {
+        if (data.hasOwnProperty(k)) {
+          data[k].key = k;
+          if (data[k].hasOwnProperty('validity')) {
+            data[k].eval = data[k].validity;
+            config.forEach(element => {
+              data[k].eval = data[k].eval.includes(element.key) ? data[k].eval
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: Looks like we are setting  `.eval` (better to rename it - validityExpression) for all config items, i.e. even if none than empty string?!

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 2]
```diff
@@ -352,4 +346,52 @@ export class ViewConfigItemComponent implements OnInit, OnChanges, AfterViewChec
...
+                .replace(new RegExp(element.key, 'g'), `'${element.value}'`) : data[k].eval;
+            });
+          }
+        }
+      }
+
+      for (const k in data) {
+        if (data.hasOwnProperty(k)) {
+          if (data[k].hasOwnProperty('validity')) {
+            try {
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: better to put try/catch thing inside if block `if (data[k]['validity'].trim() != "")`

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 3]
```diff
@@ -352,4 +346,52 @@ export class ViewConfigItemComponent implements OnInit, OnChanges, AfterViewChec
...
+              data[k].editable = eval(data[k].eval);
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: May be better `eval(data[k].validityExpression) === false ? false : true` in-case of validityExpression is incomplete e.g. "validity": "authentication" 

```
authetication="Yes"
eval("authetication")
"Yes"
```

**Review by praveen-garg — DISMISSED** [review]

**Review by praveen-garg — DISMISSED** [review]

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 4]
```diff
@@ -352,4 +340,65 @@ export class ViewConfigItemComponent implements OnInit, OnChanges, AfterViewChec
...
+
+      for (const k in data) {
+        if (data.hasOwnProperty(k)) {
+          if (data[k].hasOwnProperty('validity')) {
+            if (data[k]['validity'].trim() !== '') {
+              try {
+                // tslint:disable-next-line: no-eval
+                const e = eval(data[k].validityExpression);
+                if (typeof (e) !== 'boolean') {
+                  console.log('Validity expression', data[k].validityExpression, 'for', k, 'evlauted to non-boolean value ', e);
```
Comment (MonikaSharma06) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts:377]: Is it required to log this?

**Review by MonikaSharma06 — DISMISSED** [review]

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 5]
```diff
@@ -352,4 +342,65 @@ export class ViewConfigItemComponent implements OnInit, OnChanges, AfterViewChec
...
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts:377]: make `data[k].validityExpression` -> `data[k].validity` to log origin expression 

typo `evlauted` -> `evaluated`

also remove the extra space between `... non-boolean value ` and  `e`

**Review by praveen-garg — CHANGES_REQUESTED** [review]
validation error does not stop form to save!

**Review by MonikaSharma06 — DISMISSED** [review]

**Review by praveen-garg — APPROVED** [review]

**Pull request merged**